src/Build/Graph/ProjectGraph.cs

```diff
@@ -219,47 +218,113 @@ public ProjectGraph(string entryProjectFile)
-                ConfigurationMetadata projectToEvaluate = projectToEvaluateAndParentNode.Key;
-                ProjectGraphNode parentNode = projectToEvaluateAndParentNode.Value;
+                if (!string.IsNullOrEmpty(projectReferenceToParse.GetMetadataValue(ToolsVersionMetadataName)))
+                {
+                    throw new InvalidOperationException(string.Format(
+                        CultureInfo.InvariantCulture,
+                        ResourceUtilities.GetResourceString(
+                            "ProjectGraphDoesNotSupportProjectReferenceWithToolset"),
+                        projectReferenceToParse.EvaluatedInclude,
+                        parsedProject.Project.FullPath));
+                }
 
-                if (!_allParsedProjects.TryGetValue(projectToEvaluate, out ProjectGraphNode parsedProject))
+                string projectReferenceFullPath = projectReferenceToParse.GetMetadataValue(FullPathMetadataName);
+
+                PropertyDictionary<ProjectPropertyInstance> projectReferenceGlobalProperties = GetProjectReferenceGlobalProperties(projectReferenceToParse, globalProperties);
+                var projectReferenceConfigurationMetadata = new ConfigurationMetadata(projectReferenceFullPath, projectReferenceGlobalProperties);
+                if (nodeState.TryGetValue(projectReferenceConfigurationMetadata, out NodeState projectReferenceNodeState))
                 {
-                    parsedProject = CreateNewNode(projectToEvaluate, projectCollection);
-                    IEnumerable<ProjectItemInstance> projectReferenceItems = parsedProject.Project.GetItems(ProjectReferenceItemName);
-                    foreach (var projectReferenceToParse in projectReferenceItems)
+                    // a project reference can be in "Processed" state. If it is "InProcess" state, it is an ancestor and there is a circular dependency
+                    if (projectReferenceNodeState == NodeState.InProcess)
                     {
-                        if (!string.IsNullOrEmpty(projectReferenceToParse.GetMetadataValue(ToolsVersionMetadataName)))
+                        if (projectToEvaluate.Equals(projectReferenceConfigurationMetadata))
                         {
-                            throw new InvalidOperationException(string.Format(
-                                CultureInfo.InvariantCulture,
-                                ResourceUtilities.GetResourceString("ProjectGraphDoesNotSupportProjectReferenceWithToolset"),
-                                projectReferenceToParse.EvaluatedInclude,
-                                parsedProject.Project.FullPath));
+                            // the project being evaluated has a reference on itself
+                            throw new CircularDependencyException(string.Format(
+                                ResourceUtilities.GetResourceString("CircularDependencyInProjectGraph"),
```

Comment:
I think this should either have a dedicated "self-referencing" error or explicitly construct the error as

```
/path/to/SelfReferencing.csproj
->
/path/to/SelfReferencing.csproj
```

Just a single entry wouldn't be obvious.